Automation for Creating Polygons from Existing Points

nstclair-cc opened the pull request:
### Pull Request: Automation for Creating Polygons from Existing Points

**Pivotal Tracker Story**: [#187809206](https://www.pivotaltracker.com/story/show/187809206)

**Description**:
This PR introduces Cypress tests for automating the process of creating polygons from existing points. The tests include the following steps:
1. Add points to the table.
2. Click the "Graph IT!" button to create the polygon.
3. Wait for the Geometry tile to appear.

**Questions**:
1. Do these tests fit better in `geometry.spec.ts` or one of the table tile tests?
2. Do we want to add anything, such as resizing a polygon with the keyboard keys and confirming the resize functionality is working?

I have put in the PR to avoid further rebase issues, but please don't hesitate to ask any questions. Thank you!

---

**Changes Included**:
- Added Cypress tests for creating polygons from existing points.
- Tests include adding points to the table, clicking the "Graph IT!" button, and waiting for the Geometry tile to appear.

bgoldowsky reviewed (changes requested):
> Do these tests fit better in geometry.spec.ts or one of the table tile tests?

The test you have written most properly belongs in `geometry_table_integration_test_spec.js`, since it involves both geometry and table tiles.

> Do we want to add anything, such as resizing a polygon with the keyboard keys and confirming the resize functionality is working?

A test of keyboard functions wouldn't hurt - there isn't really a resize, but the arrow keys should work to move a point (or several points, whatever is selected). Also the delete and backspace keys should work to remove objects.

As far as testing the creation of polygons from existing points - I would take a look at `geometry_tool_spec.js` lines 280-292 which has the most basic version of this, creating a triangle from three points.  What I think would be most useful is to extend this to capture some of the related functionality:
- Create a second polygon that shares a point with the first polygon.
- After creating a polygon, click one of its vertexes to go back into editing it, add a point or two, then close it again.  Try adding both a new point and an existing point.

nstclair-cc commented:
Thanks, @bgoldowsky! I'll take a crack at it tomorrow. Those fixes make sense to me!

nstclair-cc commented:
@bgoldowsky - thank you for your feedback! I made the changes you requested, and I believe we now have coverage for what we need. However, please let me know if any additional tests should be implemented.

One thing I noticed: I added a test for "Delete point using keyboard command," but I saw that the subsequent test (labeling points) now looks for labels "BCA" instead of "ABC." Is this a bug? It doesn't seem to cause issues in the e2e specs, but it might impact undo/redo in the DOM, for example. Let me know what you think—I wasn't entirely sure what’s happening there.

Also, at the end of the circle tests, I noticed two trailing points that don’t get deleted. Is that expected behavior? Just wanted to double-check.

For the delete points test, I used undo to revert the deleted point and start fresh for the next test. I'm wondering if that's best practice since Cypress has a spec below mine that explicitly tests undo/redo.

bgoldowsky reviewed (changes requested):
This is great progress! - there are a few issues that I have noted that still need attention though.

To answer your questions:
- The change from "ABC" to "BCA" is a little mysterious. I think it may have to do with deleting one of the points and then re-creating it, the order gets altered.
- The circle test deletes a point at the end but doesn't delete all of the points - I think it's fine. It is just making sure that the circle goes away if either of the points that define it is removed.
- I think using 'undo' to reset things after a test is ok, but it can make it a little harder to read and reason about the test since exactly how much gets reverted by clicking undo once is not always obvious.

johnTcrawford reviewed (commented):
I have little to add to Boris's comments - just that one small comment I made. Aside from those suggestions, it looks good to me.

nstclair-cc commented:
In this PR, I made the following changes:

- Refined the checks related to graph point movement with keyboard strokes (following John's feedback).
- Removed the XYTile checks.
- Added a check for polygon length when points are shared.
- Improved the checks to ensure polygon lengths are as expected in the polygon with shared point checks.

For the labeling issue, I moved the test spec farther up in the sequence to prevent labels from shifting around. Once this code is merged, I'll do a quick sniff test for bugs (see [Pivotal Tracker story](https://www.pivotaltracker.com/story/show/188159177)).

I also noticed that in the Cypress test with shared points, the angle measurement seems to fluctuate. While I addressed this in the code, I’m considering removing this check, as the labels on the line segments seem to work well. I’m wondering if Cypress may have caught a bug here, so I’ll do some manual testing to determine if it’s a Cypress quirk or something more.

johnTcrawford commented on cypress/e2e/functional/tile_tests/geometry_tool_spec.js:
```diff
@@ -246,6 +254,77 @@ context('Geometry Tool', function () {
+    // Also check that the angle label has changed from its original value
+    geometryToolTile.getAngleAdornment().should(($label) => {
+        const angleText = $label.text();
+        expect(angleText).not.to.equal('90'); // 90° was the original value
```
Shouldn't this be:
expect(angleText).not.to.equal('90°');
including the degree sign (°), or am I missing something? Alternatively, you could use ".not.contain('90')".

bgoldowsky replied:
Does not appear that this was resolved.
The matching test before the point was moved (line 268) is 
   `geometryToolTile.getGraphPointLabel().contains('90°').should('exist');`
So this should be the opposite of that, eg
   `expect(angleText).not.contains('90°');`

bgoldowsky replied:
I'll go ahead and check in this fix since it was tiny & I wanted to test it anyway.

johnTcrawford commented on cypress/e2e/functional/tile_tests/geometry_tool_spec.js:
```diff
@@ -277,6 +356,164 @@ context('Geometry Tool', function () {
+      } else {
+        throw new Error(`Unexpected angle: ${angleText}`);
+      }
+    });
```
It seems odd to me that the angle is not consistent, but this is a good solution if so. And I guess it's not worth spending time investigating such a small inconsistency.

johnTcrawford commented on cypress/e2e/functional/tile_tests/geometry_tool_spec.js:
```diff
@@ -277,6 +356,164 @@ context('Geometry Tool', function () {
+      originalCx = parseFloat($point.attr('cx'));
+      originalCy = parseFloat($point.attr('cy'));
+      cy.wrap(originalCx).as('originalCx');
+      cy.wrap(originalCy).as('originalCy');
```
For a moment, this confused me: I thought that "cy.wrap(..." was referring to 'cy'. the point's y value, rather than to 'cypress'. Perhaps the 'cx' & 'cy' variable names should be changed to something like 'px' & 'py' to avoid this issue? I'm not sure what the 'c' refers to, anyway. (I guess, to be consistent, you would also need to change 'originalCx' & 'originalCy' to 'originalPx' & 'originalPy'.)

johnTcrawford reviewed (approved):
Looks good, I Approve - i.e., I don't need to review this again. However, I made several comments, of which the one about:
expect(angleText).not.to.equal('90...
(my second comment below) should at least be looked at, and changed if I am correct about it.

johnTcrawford commented on cypress/e2e/functional/tile_tests/geometry_tool_spec.js:
```diff
@@ -277,6 +356,164 @@ context('Geometry Tool', function () {
+    // Verify that the angle has decreased by 1 degree
+    cy.get('@initialAngle').then((initialAngle) => {
+      const newAngle = initialAngle - 1;
+      geometryToolTile.getGraphPointLabel().contains(`${newAngle}°`).should('exist'); // Check if angle decreased by 1 degree
```
As above for cx & cy, you could change this to check that the value is less, rather than a specific value; however, if you do, you would need to parse out (i.e. remove) the degree sign (°) from the new value returned by getGraphPointLabel(), so that you can compare integers.

bgoldowsky reviewed (approved):
This looks great!  👍
